napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,122 @@
+        # being sliced now. This allows us to slice arbitrary sets of
+        # layers with some sensible and not too complex cancellation
+        # policy.
+        existing_task = self._find_existing_task(layers)
+        # are the tasks unique?
+        if existing_task:
+            LOGGER.debug('Cancelling task for %s', layers)
+            existing_task.cancel()
+
+        # Not all layer types will initially be asynchronously sliceable.
+        # The following logic gives us a way to handle those in the short
+        # term as we develop, and also in the long term if there are cases
+        # when we want to perform sync slicing anyway.
+        requests = {}
+        for layer in layers:
+            # TODO: temporary to prove a point (which currently fails)
+            # import weakref
+            # assert isinstance(layer, weakref.ReferenceType) # this fails
+            if layer._is_async():
+                requests[layer] = layer._make_slice_request(dims)
+            else:
+                layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+        # create task for slicing of each request/layer
+        task = self._executor.submit(self._slice_layers, requests)
+
+        # construct dict of layers to layer task
+        self._layers_to_task[tuple(requests.keys())] = task
+
+        # once everything is complete, release the block
+        task.add_done_callback(self._on_slice_done)
+
+        return task
+
+    def shutdown(self) -> None:
+        """This should be called from the main thread when this is no longer needed."""
+        # TODO: kcp: doesn't this make this class instance null? Maybe we need a context manager to handle all instances off this class?
+        self._executor.shutdown()
+
+    def _slice_layers(self, requests: Dict) -> Dict:
+        """This can be called from the main or slicing thread.
+        Iterates throught a dictionary of request objects and call the slice
+        on each individual layer."""
+        return {
+            layer: layer._get_slice(request)
+            for layer, request in requests.items()
+        }
+
+    def _on_slice_done(self, task: Future[Dict]) -> None:
+        """This can be called from the main or slicing thread.
+        Release the thread.
+        This is the "done_callback" which is added to each task.
+        """
+        success = self._try_to_remove_task(task)
+
+        if not success:
+            LOGGER.debug('Task not found')
+            return
+
+        if task.cancelled():
+            LOGGER.debug('Cancelled task')
+            return
+        result = task.result()
+        self.events.ready(Event('ready', value=result))
+
+    def _try_to_remove_task(self, task):
+        """Attempt to remove task, return false if task not found, return true
+        if task removed from layers_to_task dict"""
+        with self.lock:
+            task_to_layers = {v: k for k, v in self._layers_to_task.items()}
```

Comment:
As building the inverse dict iterates over all the items, we may as well just iterate over the items to find the matching task, as it will always be less expensive.

Alternatively, you could maintain the inverse dictionary as you go.

Reply:
Fair enough, will just loop over it. Maintaining the inverse seems messier.